onnxruntime/core/providers/cpu/math/sign.cc

```diff
@@ -0,0 +1,168 @@
+
+// Unsigned Integer types
+template <class T>
+static void SignUnsigned(const Tensor* input, Tensor* output) {
+  static_assert(std::numeric_limits<T>::is_integer &&
+                    !std::numeric_limits<T>::is_signed,
+                "Expect a unsigned integer type");
+  auto span = gsl::make_span(input->Data<T>(), input->Shape().Size());
+  auto output_data = output->template MutableData<T>();
+  std::transform(span.cbegin(), span.cend(), output_data, [](T val) {
+    return (val == T(0)) ? T(0) : T(1);
+  });
+}
+
+// Signed types
+template <class T>
+void SignSignedInteger(const Tensor* input, Tensor* output) {
+  static_assert(std::numeric_limits<T>::is_integer &&
+                    std::numeric_limits<T>::is_signed,
+                "Expect a signed type");
+  auto span = gsl::make_span(input->Data<T>(), input->Shape().Size());
+  auto output_data = output->template MutableData<T>();
+  std::transform(span.cbegin(), span.cend(), output_data, [](T val) {
+    if (val > T(0)) {
+      return T(1);
+    } else if (val < T(0)) {
+      return T(-1);
+    }
+    return T(0);
+  });
+}
+
+// The spec does not specify how NaN is
+// treated but we have to treat it somehow. We choose
+// to return 0 for NaN as TF does.
+template <class T>
+inline T FloatingImpl(T val) {
+  if (std::isnan(val) || val == T(0)) {
+    return T(0);
+  } else if (val > T(0)) {
+    return T(1);
+  } else {
+    return T(-1);
+  }
+}
+
+template <class T>
+void SignFloat(const Tensor* input, Tensor* output) {
+  static_assert((std::is_same<T, float>::value || std::is_same<T, double>::value),
+                "Expect a signed type");
+  auto span = gsl::make_span(input->Data<T>(), input->Shape().Size());
+  auto output_data = output->template MutableData<T>();
+  std::transform(span.cbegin(), span.cend(), output_data, [](T val) {
+    return FloatingImpl<T>(val);
+  });
+}
+
+void SignMLFloat16(const Tensor* input, Tensor* output) {
+  auto span = gsl::make_span(input->Data<MLFloat16>(), input->Shape().Size());
+  auto output_data = output->template MutableData<MLFloat16>();
+  std::transform(span.cbegin(), span.cend(), output_data, [](const MLFloat16& val) {
+    float fl = math::halfToFloat(val.val);
+    return MLFloat16(math::floatToHalf(FloatingImpl(fl)));
+  });
+}
+
+void SignBFloat16(const Tensor* input, Tensor* output) {
+  auto span = gsl::make_span(input->Data<BFloat16>(), input->Shape().Size());
+  auto output_data = output->template MutableData<BFloat16>();
+  std::transform(span.cbegin(), span.cend(), output_data, [](const BFloat16& val) {
+    float fl = val.ToFloat();
+    return BFloat16(FloatingImpl(fl));
+  });
+}
+
+}  // namespace sign_internal
+
+Status Sign::Compute(OpKernelContext* ctx) const {
+  using namespace sign_internal;
+
+  auto input = ctx->Input<Tensor>(0);
+  auto output = ctx->Output(0, input->Shape());
+
+  auto dtype = input->DataType();
+  if (dtype == DataTypeImpl::GetType<int8_t>()) {
```

Comment:
use switch ?

Reply:
No. Switch won't work for pointers?

Reply:
C++ `switch` does not operator on pointers and `DataType` returns an MLDataType that is a NON_ONNX primitive that does not have a TypeProto associated with it so TensorProto_DataType enum value can not be easily obtained.

Reply:
I understand you can't directly use the input->DataType however there is an utility method already available, you can check this out utils::GetTensorProtoType… This is not a must do ….

Reply:
Discussed offline with the utility function doing almost the same thing.